게임 데이터 저장소 추가 + 기타 리팩토링 진행 중

yokxim2 opened the pull request:
## ✨ 담당 파트
- 스프링 이벤트 적용 전 의존성 분리 작업 중

## 🔎 작업 상세 내용
- 커밋 메시지를 참고 바랍니다.

## 🔧 앞으로의 과제
- 단기 목표: 기존에 있던 AccountGameData 객체를 최근 변경 날짜 비교해서 수정(이전 날짜부터 전적 조회 추가, 올드한 전적 삭제) 로직을 구현해야 합니다.

## ✅ 테스트 코드 작성 및 기능 테스트 여부
- [] 테스트 코드 작성
- [] 기능 테스트 여부

## ➕ 이슈 링크
- #30

yokxim2 committed:
feat(account-gameData) 계정의 모든 게임 정보를 연결하는 DB Entity

yokxim2 committed:
fix(duoService) DuoService 리팩토링

RiotClientService에 대한 의존성을 낮추는 작업 중입니다.

yokxim2 committed:
fix(Favorite) Favorite 필드 인덱싱 추가 #30

- 조회 성능 향상을 위해 accountId, queueType, championName에 인덱싱을 추가했습니다.
- playCount, winCount에 기본값을 추가했습니다.

yokxim2 committed:
feat(flexRankData) 자유랭크 게임 데이터 Entity #30

- id, tier, ranks, wins, totalGames, Kda
- 최근 선호 챔피언은 Favorite 으로 관리하려고 합니다.

yokxim2 committed:
feat(quick-gameData) 일반 게임 데이터 Entity #30

- id, wins, totalGames, Kda
- 최근 선호 챔피언은 Favorite 으로 관리하려고 합니다.

yokxim2 committed:
feat(soloRank-data) 솔로랭크 게임 데이터 Entity #30

- id, tier, ranks, wins, totalGames, Kda
- 최근 선호 챔피언은 Favorite 으로 관리하려고 합니다.

yokxim2 committed:
feat(time-util) convertToEpochSeconds 메서드 #30

- LocalDateTime을 입력값으로 줄 때 epoch 단위시간을 반환하는 메서드입니다.
- AccountGameData는 업데이트한 날짜를 LocalDateTime 타입으로 같이 저장하므로, 해당 날짜부터 현재 날짜까지 RiotClient의 getMatchIds 메서드를 호출하기 쉽게 LocalDateTime 시간값을 변환하는데 도움을 줄 메서드입니다.
- 이 메서드가 불필요해 보인다면 꼭 의견 주십시오.

yokxim2 committed:
feat(riot-data-provider) RiotClientService 부모 인터페이스 #30

객체 상호 간 의존성을 낮추기 위해 만들어 본 인터페이스입니다.

yokxim2 committed:
chore(riot-client-service) 인터페이스 구현 #30

yokxim2 committed:
fix(account) Account -> AccountGameData 단방향 의존성 #30

Account를 먼저 조회한 후 해당 계정의 전적을 검색하는 로직이 많을 것 같아 Account에서 AccountGameData를 바라보게 바꿨습니다.

yokxim2 committed:
fix(account-detail) profileIconId 추가 #30

profileIconId를 accountGameId에 넣기에 너무 이질적일 것 같아서, 계정 상세정보를 저장하는 AccountDetail에 추가했습니다.

yokxim2 committed:
refactor(account-game-data) 팩토리 메서드 및 update 메서드 추가 #30

yokxim2 committed:
feat(account-game-data-repository) JpaRepository #30

yokxim2 committed:
feat(flex-rank-data-repository) JpaRepository #30

yokxim2 committed:
feat(quick-game-data-repository) JpaRepository #30

yokxim2 committed:
feat(solo-rank-data-repository) JpaRepository #30

yokxim2 committed:
feat(queue-data) duo 업데이트를 위한 값들을 모아놓은 DTO #30

티어, 랭크, 승리 횟수, 패배 횟수, kda를 모아 놓은 DTO입니다. 이 DTO가 불필요하다고 생각되시면 꼭 말씀해 주십시오.

yokxim2 committed:
refactor(account-service) 리팩토링 #30

Account 객체 생성을 위해서는 AccountGameData를 불러와야 하는데, 이후 개발할 retrieveData 메서드가 바로 이벤트를 발행할 메서드입니다. 추후 이벤트 리스너 메서드와 함께 개발하도록 하겠습니다.

yokxim2 committed:
refactor(duo-service) rebase #30

DuoService에서 RiotClient 관련 모든 의존성을 제거하고, Duo 객체만을 만들고, 수정하고, 삭제하는 책임만을 가지도록 수정했습니다. 반복 코드를 최대한 줄이려고 노력했으나, 아직 개선이 더 필요하다 생각하시면 꼭 말씀해 주세요.

yokxim2 committed:
refactor(riot-client-service) 리팩토링 #30

Account 객체를 가져오게 하지 않고, 값을 풀어서 가져오도록 하여 의존성을 지웠습니다.

yokxim2 committed:
refactor(riot-data-provider) 리팩토링 #30

RiotClientService 내부로직 수정으로 인수 값을 수정했습니다.

for2ver0 commented on src/main/java/com/summoner/lolhaeduo/common/util/TimeUtil.java:
```diff
@@ -3,14 +3,23 @@
 
-    public long startTimeInEpoch(int days) {
+    public static long startTimeInEpoch(int days) {
         Instant now = Instant.now();
```
- convertToEpochSeconds 메서드에서 UTC 사용하셨던데, 일관성을 위해서 다음 코드와 같이 ZoneOffset.UTC 를 명시적으로 표현하면 어떨까요?
`Instant now = Instant.now(Clock.systemUTC());`

for2ver0 commented on src/main/java/com/summoner/lolhaeduo/common/util/TimeUtil.java:
```diff
@@ -3,14 +3,23 @@
 public class TimeUtil {
 
-    public long startTimeInEpoch(int days) {
+    public static long startTimeInEpoch(int days) {
```
- convertToEpochSeconds 메서드에서 매개변수에 대한 검증? 부분도 추가해주셨던데, 이 메서드에도 추가하면 어떨까요?
```Java
if (days < 0) {
     throw new IllegalArgumentException("시작 일자는 미래가 될 수 없습니다.");
}
```

yokxim2 replied:
네 추가했습니다!

for2ver0 commented on src/main/java/com/summoner/lolhaeduo/domain/account/entity/AccountGameData.java:
```diff
@@ -0,0 +1,34 @@
+    @OneToOne(fetch = FetchType.LAZY)
+    private FlexRankData flexRankData;
+
+    private LocalDateTime lastUpdated;
```
AccountGameData는 Timestamped 인터페이스를 상속받지 않고 최근 수정 날짜만 저장한 이유가 따로 있을까요?

yokxim2 replied:
듣고 보니 그렇네요. 수정하겠습니다~

for2ver0 commented on src/main/java/com/summoner/lolhaeduo/domain/duo/dto/QueueData.java:
```diff
@@ -0,0 +1,25 @@
+import lombok.Getter;
+
+@Getter
+public class QueueData {
```
빠른 대전은 tier와 rank가 솔로 랭크로부터 반환받기도 하고.. 뭔가 QueueData라는 이름이 멤버 변수들과 매칭이 잘 되지 않아서 저는 어색하네요..
좀 더 적절한 네이밍이 생각나면 추후에 변경하면 좋겠습니다~

yokxim2 replied:
오케이 추후에 고려해보도록 하죠!

for2ver0 commented on src/main/java/com/summoner/lolhaeduo/domain/account/service/AccountService.java:
```diff
@@ -63,10 +68,16 @@ public void linkAccount(LinkAccountRequest request, Long memberId) {
         }
     }
+
+    // 초기 데이터 불러오기 (추후 비동기 처리로 개발 예정)
```
이런 코멘트 감사합니다 :)

for2ver0 commented on src/main/java/com/summoner/lolhaeduo/domain/duo/service/DuoService.java:
```diff
@@ -43,142 +37,87 @@ public DuoCreateResponse createDuo(DuoCreateRequest request, Long memberId) {
         );
 
-        if (linkedAccount.getMemberId() != memberId) {
+        if (linkedAccount.getMemberId().equals(memberId)) {
```
앗!! `!`가 빠진 것 같습니다!
```suggestion
        if (!linkedAccount.getMemberId().equals(memberId)) {
```

yokxim2 replied:
아 맞네요. 감사합니다~

for2ver0 reviewed (commented):
리뷰 확인 부탁드립니다~

yokxim2 committed:
test(riot-client-service-test) 테스트 잠시 삭제 #30

yokxim2 committed:
chore(riot-data-provider) getProfileIconUrl 메서드 삭제 #30

인터페이스로서 공유되지 않을 메서드라고 판단해서 지웠습니다.

yokxim2 committed:
chore(time-util) startTimeInEpoch 삭제 #30

몇일 전 시간을 인수 값으로 넣는 것보다, LocalDateTime 타입 값을 인수로 받는 것이 더 확장성이 높다고 판단하여 수정했습니다.

yokxim2 committed:
feat(solo-rank-data) 정적 팩토리 메서드 추가 #30

yokxim2 committed:
feat(quick-game-data) 정적 팩토리 메서드 추가 #30

yokxim2 committed:
feat(flex-rank-data) 정적 팩토리 메서드 추가 #30

yokxim2 committed:
feat(async-config) 이벤트 비동기 호출을 위한 config #30

@EnableAsync를 활용해서 이벤트 비동기 호출을 가능하게 했습니다.

yokxim2 committed:
feat(account-game-date-event) 계정 연동 시 발생하는 이벤트 #30

계정을 연동하면 실행되는 이벤트입니다. 계정에 관련된 모든 정보를 가져옵니다.

yokxim2 committed:
feat(account-game-data) 영속성 전이 추가 #30

AccountGameData 객체를 생성, 수정, 삭제하면 하위 항목인 SoloRankData, FlexRankData, QuickGameData가 동시에 반영될 수 있도록 연관관계를 매핑했습니다.

yokxim2 committed:
fix(account-detail) profileIconId 삭제 #30

profileIconId는 업데이트 해야 할 항목이므로, AccountDetail에서 사용하기 어려워서 직접 AccountGameDataService에서 다루도록 했습니다.

yokxim2 committed:
feat(account) AccountGameData 항목 연결 #30

AccountGameData는 Account 객체가 생성되기 전에 만들어서 생성자로 전달되지 않습니다. 추후 AccountGameDataService에서 AccountGameData 객체를 만들고 난 후, linkAccountGameData를 호출해서 값을 주입해줍니다.

yokxim2 committed:
fix(account-service) riotClient 의존성 삭제 및 이벤트 추가 #30

- RiotClient를 오직 RiotClientService 내에서만 호출하도록 리팩토링했습니다. AccountService에서는 앞으로 Account를 생성하는 로직만 있습니다.
- 계정을 연동하고 나면, AccountGameDataEvent 이벤트를 발생시킵니다.

yokxim2 committed:
fix(riot-client-service) 내부 로직 수정 #30

- AccountService에 있던 RiotClient를 호출하는 로직을 RiotClientService 내부에서 호출하도록 리팩토링했습니다.
- 프로필 아이콘 URL은 RiotClient를 호출하는 로직이기에, 이곳에 만들어주는게 맞다고 판단했습니다.
- TimeUtil의 startTimeInEpoch를 convertToEpochSeconds로 대체했습니다.

yokxim2 committed:
feat(account-game-data-service) AccountGameData 서비스 로직 #30

- AccountGameData를 만드는 로직입니다. createAccountGameDataEvent는 AccountGameDataEvent가 발행되면, 동작을 시작하는 메서드입니다.
- RiotClientService의 로직을 사용합니다. AccountGameData를 생성하면, Account 내 linkAccountGameData 메서드를 사용하여 새로 만든 AccountGameData를 기존의 Account에 주입합니다.
- 생성은 구현했으나, 아직 수정을 구현하지 않았습니다. 추후 updatedAt을 비교해서 지난번 수정 이후 ~ 현재까지 전적을 조회해서 기존 AccountGameData에 추가하고, 너무 오래된 Data는 삭제할 수 있도록 구현하겠습니다.

for2ver0 reviewed (commented):
레깅스님 너무 수고하셨습니다 👍 👍 👍

yokxim2 committed:
chore(duo-service) 불필요한 import 삭제 #30

yokxim2 committed:
fix(account-service) 이벤트 코드 순서 변경 #30

이벤트 발행 시 AccountId를 전달해주기 때문에, Account를 save한 후 이벤트가 발생될 수 있도록 수정했습니다.

yokxim2 committed:
fix(account-game-data-service) 외부 API 호출 인자 수정 #30

getRankGameStats에는 accountId 대신 SummonerId가 필요했기 때문에 수정합니다.

for2ver0 reviewed (approved):
수고하셨습니다 👍 👍 👍

hubaek commented on src/main/java/com/summoner/lolhaeduo/domain/account/service/AccountGameDataService.java:
```diff
@@ -0,0 +1,83 @@
+        // 매치 아이디 가져오기
+        List<String> quickMatchIds = riotClientService.getMatchIds(QUICK, NUMBER_OF_RECENT_MATCH, account.getRegion(), account.getAccountDetail().getPuuid());
+        List<String> soloMatchIds = riotClientService.getMatchIds(SOLO, rankStats.getSoloTotalGames(), account.getRegion(), account.getAccountDetail().getPuuid());
+        List<String> flexMatchIds = riotClientService.getMatchIds(FLEX, rankStats.getFlexTotalGames(), account.getRegion(), account.getAccountDetail().getPuuid());
```
결국 3가지로 가져올 수 밖에 없는거군요 ㅠ

yokxim2 replied:
ㅋㅋ.. 이게 제가 생각 가능한 유일한 방법이었슴다

hubaek reviewed (approved):
LGTM 굉장하군요
